crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -0,0 +1,389 @@
+use matrix_sdk_base::deserialized_responses::TimelineEvent;
+use matrix_sdk_test::{
+    test_json, InvitedRoomBuilder, JoinedRoomBuilder, KnockedRoomBuilder, LeftRoomBuilder,
+    SyncResponseBuilder,
+};
+use ruma::{OwnedEventId, OwnedRoomId, RoomId};
+use serde_json::json;
+use wiremock::{
+    matchers::{header, method, path, path_regex},
+    Mock, MockBuilder, MockGuard, MockServer, Respond, ResponseTemplate, Times,
+};
+
+use super::logged_in_client;
+use crate::{Client, Room};
+
+pub struct MatrixMockServer {
+    server: MockServer,
+    client: Client,
+
+    /// Make the sync response builder stateful, to keep in memory the batch
+    /// token and avoid the client ignoring subsequent responses after the first
+    /// one.
+    sync_response_builder: Arc<Mutex<SyncResponseBuilder>>,
+}
+
+impl MatrixMockServer {
+    /// Create a new mocked server specialized for Matrix usage.
+    pub async fn new() -> Self {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri().to_string())).await;
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Creates a new [`MatrixMockServer`] when both parts have been already
+    /// created.
+    pub fn from_pair(server: MockServer, client: Client) -> Self {
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Return the underlying client.
+    pub fn client(&self) -> Client {
+        self.client.clone()
+    }
+
+    /// Return the underlying server.
+    pub fn server(&self) -> &MockServer {
+        &self.server
+    }
+
+    /// Overrides the sync/ endpoint with knowledge that the given
+    /// invited/joined/knocked/left room exists, runs a sync and returns the
+    /// given room.
+    pub async fn sync_room(&self, room_id: &RoomId, room_data: impl Into<AnyRoomBuilder>) -> Room {
+        let any_room = room_data.into();
+
+        self.mock_sync()
+            .ok_and_run(move |builder| match any_room {
+                AnyRoomBuilder::Invited(invited) => {
+                    builder.add_invited_room(invited);
+                }
+                AnyRoomBuilder::Joined(joined) => {
+                    builder.add_joined_room(joined);
+                }
+                AnyRoomBuilder::Left(left) => {
+                    builder.add_left_room(left);
+                }
+                AnyRoomBuilder::Knocked(knocked) => {
+                    builder.add_knocked_room(knocked);
+                }
+            })
+            .await;
+
+        self.client.get_room(room_id).expect("look at me, the room is known now")
+    }
+
+    /// Overrides the sync/ endpoint with knowledge that the given room exists
+    /// in the joined state, runs a sync and returns the given room.
+    pub async fn sync_joined_room(&self, room_id: &RoomId) -> Room {
+        self.sync_room(room_id, JoinedRoomBuilder::new(room_id)).await
+    }
+
+    pub async fn verify_and_reset(&self) {
```

Comment:
I personally have no idea what these methods do, so would be nice to have a comment.